More documentation

nical opened the pull request:
This is a work in progress. This PR adds the gfx::doc::terminology module which only contains rustdoc documentation. The idea is to have a centralized place with information about what various gfx concepts map to in other APIs. I am not quite certain what the best way to present this information is. For example we could also have sections "OpenGL to gfx", "D3D11 to gfx", etc which would only link to the documentation of each term in gfx concisely (like: " - GL thing: [gfx thing](link to thing in the doc)".

What I am trying to address here is that for someone who does not know the internals of gfx, using this crate currently is a small exercise in reverse-engineering, especially when you know what you want to do (but not how it is called in gfx).

The other thing that I want to move towards, is for each module (or at least each crate) to have a large doc block with sections `Overview` and `Examples` that give some info about the general design and what the most important concepts are with link to teh relevant pieces. Rustoc is not good at making the important things stand out, so it has to be done somewhat manually in the crate/module doc.

Ultimately, gfx should strive to get to a documentation at the level of [glium's](https://docs.rs/glium) which I took a lot of inspiration from when documenting [lyon](https://docs.rs/lyon). I received some positive feedback from the work on lyon and I think that gfx could really use some doc that is structured similarly.

I don't think should land as is with all of the TODO's. Don't hesitate to review this while the work is in progress to help me fill the holes and rewrite/rearrange as needed.

nical commented:
It might be better to make the terminology page just a list of links to various areas of the documentation and move the extra explanation closer to their types in the doc.

msiglreith commented on src/core/src/factory.rs:
```diff
@@ -12,10 +12,41 @@
+//!
+//! ## Unordered access
+//!
+//! Unordered access views refer to buffers that can be written to by a shader. 
```
We should annotate here that they support read and write to avoid confusion.

kvark replied:
UAVs don't refer to just buffers, also can view into textures

msiglreith commented on src/render/src/doc/terminology.rs:
```diff
@@ -0,0 +1,49 @@
+// Copyright 2016 The Gfx-rs Developers.
```
nit: year

msiglreith commented on src/render/src/doc/terminology.rs:
```diff
@@ -0,0 +1,49 @@
+//!
+//! ## Slice
+//!
+//! A `Slice` in gfx describes for a givien draw call the range of vertices to process, their order
```
typo: `givien`/`given`

msiglreith commented on src/render/src/lib.rs:
```diff
@@ -14,8 +14,63 @@
+//!
+//! ## Command buffers and encoders
+//!
+//! A command buffer is an already built serialized list drawing and compute commands.
```
This sentence is a bit confusing to me, is there some word missing?

jonas-schievink commented on src/render/src/lib.rs:
```diff
@@ -14,8 +14,63 @@
+//! ## Command buffers and encoders
+//!
+//! A command buffer is an already built serialized list drawing and compute commands.
+//! Unlike with vulkan, command buffers are what you use to create commands, but only
```
Is there a "not" missing in this sentence?

jonas-schievink commented on src/render/src/lib.rs:
```diff
@@ -14,8 +14,63 @@
+//! Each gfx backend provides its own factory type which implements both:
+//!
+//! - The [`factory` trait documentation](traits/trait.Factory.html#overview).
+//! - The [`factoryExt` trait documentation](traits/trait.FactoryExt.html).
```
"documentation" seems out of place here. Also, traits should be capitalized correctly.

kvark commented on src/render/src/doc/terminology.rs:
```diff
@@ -0,0 +1,49 @@
+//!
+//! ## Handle
+//!
+//! Handles are the mechanism primarily gfx uses to refer to most GPU resources.
```
"gfx primarily" - a swap of these sounds better to me

kvark commented on src/render/src/doc/terminology.rs:
```diff
@@ -0,0 +1,49 @@
+//! In OpenGL, handles are simple integers (contrary to gfx's high level API, OpenGL does not
+//! provide any automatic memory management).
+//!
+//! Note: The comments about memory management above apply to the high level API, but not the
```
I see "high level" here being a confusing/over-loaded term

nical replied:
Agreed. How would you name the high level api as opposed to the ll one?

kvark commented on src/render/src/doc/terminology.rs:
```diff
@@ -0,0 +1,49 @@
+//! and the number of instances.
+//! The slice is how index buffers are specfied in a draw call.
+//!
+//! Not to confuse with the term slice as employed more broadly in rust (`&[T]`) which is a view into
```
`Slice` is just not a very good name for our concept...

nical replied:
I agree. It confused me until I looked into its code, and the notion of slice is so central in rust that I think it would be good to avoid overloading it.

kvark commented on src/render/src/lib.rs:
```diff
@@ -14,8 +14,63 @@
+//! - The [`factoryExt` trait documentation](traits/trait.FactoryExt.html).
+//!
+//! OpenGL does not have a notion of factory (resources are created directly off of the global
+//! context). D3D11 has a DXGI factroy but it is only used to interface with other processes
```
"factroy"

kvark commented on src/render/src/lib.rs:
```diff
@@ -14,8 +14,63 @@
+//!
+//! OpenGL does not have a notion of factory (resources are created directly off of the global
+//! context). D3D11 has a DXGI factroy but it is only used to interface with other processes
+//! and the window manager, resources like textures are usually created using the device.
```
In fact `gfx::Factory` is roughly equivalent to `ID3D11Device`, `MTLDevice` and `VkDevice`.
@msiglreith - we should consider renaming it to `Device`. Even if confusing, at least it goes in line with the underlying backends

msiglreith replied:
Sounds reasonable. Factory interacts now with fences therefore `Device` would be more suitable. We need to find a new name for our `Device` struct then!

nical commented:
I see that Factory has ` *_raw` methods for untyped buffer creation and `*_u8` for textures. Do they mean different things?

kvark commented:
@nical 
> Agreed. How would you name the high level api as opposed to the ll one?

Maybe "current core" as opposed to "low level core"? I don't see a good name :(

> I see that Factory has *_raw methods for untyped buffer creation and *_u8 for textures. Do they mean different things?

`_u8` has been added on top, it's not soundly matching the existing APIs. It still operates on typed handles, just expects byte slices of data.

@msiglreith 
> We need to find a new name for our Device struct then!

If you are referring to CoreLL `Device`, then it looks rather auxiliary, so the name doesn't matter much. Could be `ActiveAdapter` or something.

msiglreith commented on src/core/src/factory.rs:
```diff
@@ -187,17 +187,52 @@ pub enum WaitFor {
+/// The term "raw" is used in the context of types of functions that have a strongly typed and an
+/// untyped equivalent, to refer to the untyped equivalent.
+///
+/// For example ['Encloder::create_buffer_raw'](trait.Factory.html#tymethod.create_buffer_raw) and
```
`Encloder`?

msiglreith commented on src/render/src/lib.rs:
```diff
@@ -14,8 +14,79 @@
+//!
+//! See:
+//!
+//! - The [`Encoder` struct documentation](struct.Encoder.html).
```
We only expose `GraphicsEncoder` atm (after the ll integration got merged, the old `Encoder` got renamed)

msiglreith commented on src/render/src/lib.rs:
```diff
@@ -14,8 +14,79 @@
+//!
+//! ## Device
+//!
+//! See [the `gfx::Device` trait](trait.Device.html).
```
`Device` is now a struct in `core`

msiglreith commented on src/render/src/lib.rs:
```diff
@@ -14,8 +14,79 @@
+//!
+//! Handles internally use atomically reference counted pointers to deal with memory management.
+//! GPU resources are not destroyed right away when all references to them are gone. Instead they
+//! are destroyed the next time [Device::cleanup](trait.Device.html#tymethod.cleanup) is called.
```
`cleanup` is now a method of command queues

nical commented:
> A few things changed with the first ll integration (sorry for the extra work!). Would you mind addressing the points I mentioned?

No problem, on it.
